lib/rom/commands/result.rb

```diff
@@ -24,6 +24,20 @@ def to_ary
       end
       alias_method :to_a, :to_ary
 
+      # Return true if command successful
+      #
+      # @api public
+      def success?
+        !value.nil? && error.nil?
+      end
+
+      # Return true if command failed
+      #
+      # @api public
+      def failure?
+        value.nil? && !error.nil?
```

Comment:
So looks like we'll only ever instantiate "Success" or "Error" classes, so you could check EITHER @value OR @error and it would be sufficient. Although nothing stops someone going Result.new and having both nil, it's pretty unrealistic usage!

 ...it feels like a haskell GADT with a (not quite) unified object interface. In Haskell you would have a type that can be one or the other, and nothing else -- and you'd have to case-match on the type before being able to call "value" or "error".

WDYT about going type-based and on what you initialized rather than trying to infer:

    def success?
      is_a?(Success)
    end

I guess it depends on whether this is open for new classes to be built that inherit Result (oo style). Then this might be bad, but seems like this is going functional/type style (it's an Either).

Another alternative:

    class Success
      def success?
        true
      end
    ...
    class Error
      def success?
        false
      end
    ...